File: php/ibroadcast-uploader.php
<?php
require 'vendor/autoload.php';

use chillerlan\QRCode\{QRCode, QROptions};
use chillerlan\QRCode\Output\QROutputInterface;
use chillerlan\QRCode\Output\QRString;
use chillerlan\QRCode\Data\QRMatrix;

const CLIENT_ID = 'de4ce836a9fb11f0bc7fb49691aa2236';
const USER_AGENT = 'php uploader 1.0';

ini_set('memory_limit' , '1024M');

class OAuthError extends Exception {
    public $error_code;

    public function __construct(string $errorCode, string $message, int $code = 0, ?Throwable $previous = null) {
        $this->error_code = $errorCode;
        parent::__construct($message, $code, $previous);
    }

    public function __toString(): string {
        return "Authorization error [{$this->error_code}]: {$this->getMessage()}";
    }
}

class HTTPError extends Exception {
    public $status_code;

    public function __construct(string $statusCode, string $message, int $code = 0, ?Throwable $previous = null) {
        $this->status_code = $statusCode;
        parent::__construct($message, $code, $previous);
    }

    public function __toString(): string {
        return "HTTP error [{$this->error_code}]: {$this->getMessage()}";
    }
}

$token = loadToken();

$token = login($token);

if ($token === null) {
	fwrite(STDOUT, 'Unable to log in.' . "\n");
	exit(0);
}

saveToken($token);

// Fetch supported types
fwrite(STDOUT, 'Fetching account info...' . "\n");

$serviceContent  = getSupportedTypes($token);

if ($serviceContent == null || !isset($serviceContent->user)) {
	fwrite(STDOUT, ($serviceContent ? $serviceContent->message : 'Unable to get supported types.') . "\n");
	exit(0);
}

// Extract supported file formats from the service response
$supportedFormat = getSupportedFormats($serviceContent);

fwrite(STDOUT, 'Searching for files...' . "\n");

// Fetch the file list from the service
$syncContent     = getSyncContent($token);
$alreadyUploaded = isset($syncContent->md5) && is_array($syncContent->md5) ? $syncContent->md5 : array();

$files = array();

// Bootstrap the script, recursive method in order to traverse sub-folders
readFiles(getcwd(), $files);

// Display a nice welcome text with instructions
fwrite(STDOUT, sprintf("Found %s file(s). Press \"L\" to list or \"U\" to start the upload. [Q to Quit]\n", count($files)));
do {
	do {
		// Wait for user input
		$cmd = strtoupper(fgetc(STDIN));

	} while (trim($cmd) == '');

	if ($cmd == 'L') {

		// Display a list of files that were found in the dir and all child-dirs
		cmdList($files);
	} else if ($cmd == 'U') {

		// Trigger the upload action
		cmdUpload($files, $alreadyUploaded, $token);
		exit(0);
	}

// Allow user to abort using "Q"
} while ($cmd !== 'Q');


function getDefaultPostData($mode) {
	return array(
		'mode' => $mode,
		'client' => 'php uploader',
		'version' => '1.0',
		'device_name' => 'php uploader',
		'user_agent' => USER_AGENT
	);
}

function cmdList($files) {
	fwrite(STDOUT, "Listing found, supported files\n");

	// Display all files that are available for uploading
	foreach ($files as $file) {
		fwrite(STDOUT, $file . "\n");
	}

	fwrite(STDOUT, "Press 'U' to start the upload if this looks reasonable\n");
}

function cmdUpload($files, $alreadyUploaded, $token) {
	fwrite(STDOUT, "Uploading..\n");

	foreach ($files as $file) {
		fwrite(STDOUT, "Uploading: $file\n");

		$checksum = md5_file($file);

		// Check if the file already exists on the service, if yes, then skipp
		if (in_array($checksum, $alreadyUploaded)) {
			fwrite(STDOUT, "  skipping, already uploaded.\n");
			continue;
		}

		// File is allowed to be uploaded
		$success = false;
		try {
			$success = uploadFile($file, $token);
		} catch (HTTPError $e) {
			if ($e.status_code == "401") {
				$token = refreshTokenIfNecessary();
				$success = uploadFile($file, $token);
			}
		}

		if ($success) {
			fwrite(STDOUT, " Done!\n");
		} else {
			fwrite(STDOUT, "Failed!\n");
		}
	}
}

/**
 * Loads the token from the JSON file.
 */
function loadToken() {
	$path = __DIR__ . '/ibroadcast-uploader.json';

	if (!file_exists($path)) {
		return null;
	}

	try {
		$jsonContents = file_get_contents($path);

		$data = json_decode($jsonContents, true, 512, JSON_THROW_ON_ERROR);

		if (isset($data['token'])) {
			return $data['token'];
		}
	} catch (Throwable $e) {
		// Silently fail if file is unreadable or corrupt, as in the Python original.
	}

	return null;
}

/**
 * Saves the current token to the JSON file.
 */
function saveToken($token) {
	$path = __DIR__ . '/ibroadcast-uploader.json';

	try {
		$data = ['token' => $token];
		file_put_contents($path, json_encode($data, JSON_THROW_ON_ERROR));
		return true;
	} catch (Throwable $e) {
		fwrite(STDOUT, 'Warning, unable to save token to ibroadcast-uploader.json: ' . $e->getMessage() . PHP_EOL);
		return false;
	}
}

/**
 * The main login flow using the OAuth Device Code grant.
 * @return bool True on success, false on failure.
 */
function login($token) {
	$deviceCode = null;

	// Attempt to refresh the token if it's expired
	$token = refreshTokenIfNecessary($token);

	// Loop until we have a valid token
	while ($token === null) {
		// Get a new device code if we don't have one
		if ($deviceCode === null) {
			try {
				$deviceCode = oauthDeviceCode();
				$deviceCode['expires_at'] = time() + $deviceCode['expires_in'];

				// Generate and show QR code for the terminal
				$options = new QROptions(['outputType' => QROutputInterface::STRING_TEXT]);
				$qrcode = (new QRCode($options))->render($deviceCode['verification_uri_complete']);
				fwrite(STDOUT, $qrcode . PHP_EOL);

				fwrite(STDOUT, "To authorize, scan the QR code or enter code {$deviceCode['user_code']} at: {$deviceCode['verification_uri']}" . PHP_EOL);
				fwrite(STDOUT, PHP_EOL . 'Waiting for authorization...' . PHP_EOL);
			} catch (Throwable $e) {
				fwrite(STDOUT, 'Unable to get device code: ' . $e->getMessage() . PHP_EOL);
				return null;
			}
		}

		// Check if the device code has expired
		if ($deviceCode['expires_at'] <= time()) {
			fwrite(STDOUT, 'Device code timed out!' . PHP_EOL);
			$deviceCode = null; // Remove it to start over
			continue;
		}

		// Poll for the authorization token
		try {
			$token = oauthToken($deviceCode['device_code']);
			$token['expires_at'] = time() + $token['expires_in'];
		} catch (OAuthError $e) {
			if ($e->error_code === 'authorization_pending') {
				sleep($deviceCode['interval']); // Wait and poll again
				continue;
			}
			fwrite(STDOUT, 'Authorization error: ' . $e->getMessage() . PHP_EOL);
			return null;
		}

		// Authorization successful! Save the token and break the loop.
		saveToken($token);
		break;
	}

	return $token;
}

/**
 * Checks if the token is expired and refreshes it if necessary.
 */
function refreshTokenIfNecessary($token) {
	if ($token === null || !isset($token['expires_at'])) {
		return null;
	}

	if ($token['expires_at'] <= time()) {
		try {
			$token = refreshToken($token['refresh_token']);
			$token['expires_at'] = time() + $token['expires_in'];
			saveToken(); // Save the newly refreshed token
		} catch (OAuthError $e) {
			fwrite(STDOUT, 'Authorization error, please log in again: ' . $e->getMessage() . PHP_EOL);
			$token = null; // Invalidate the token
			saveToken(); // Save the invalidated state
		}
	}

	return $token;
}

/**
 * Gets a device code from the OAuth server.
 * @throws OAuthError
 * @return array
 */
function oauthDeviceCode() {
	fwrite(STDOUT, 'Getting device code...' . PHP_EOL);

	$queryParams = http_build_query([
		'client_id' => CLIENT_ID,
		'scope' => 'user.account:read user.upload'
	]);
	$url = 'https://oauth.ibroadcast.com/device/code?' . $queryParams;

	return makeRequest($url, 'GET');
}

/**
 * Gets an access token given a device code.
 * @param string $code The device code.
 * @throws OAuthError
 * @return array
 */
function oauthToken($code) {
	$body = [
		'client_id' => CLIENT_ID,
		'grant_type' => 'device_code',
		'device_code' => $code
	];

	return makeRequest('https://oauth.ibroadcast.com/token', 'POST', $body);
}

/**
 * Refreshes an access token using a refresh token.
 * @param string $refreshToken
 * @throws OAuthError
 * @return array
 */
function refreshToken($refreshToken) {
	fwrite(STDOUT, 'Refreshing token...' . PHP_EOL);
	$body = [
		'client_id' => CLIENT_ID,
		'grant_type' => 'refresh_token',
		'refresh_token' => $refreshToken
	];

	return makeRequest('https://oauth.ibroadcast.com/token', 'POST', $body);
}

/**
 * A helper function to make cURL requests.
 *
 * @param string $url The URL to request.
 * @param string $method The HTTP method (GET or POST).
 * @param array|null $postData Data for POST requests.
 * @return array The JSON-decoded response.
 * @throws OAuthError|JsonException
 */
function makeRequest($url, $method = 'GET', $postData = null) {
	$ch = curl_init();

	curl_setopt($ch, CURLOPT_URL, $url);
	curl_setopt($ch, CURLOPT_USERAGENT, USER_AGENT);
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

	if ($method === 'POST') {
		curl_setopt($ch, CURLOPT_POST, true);
		curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($postData));
	}

	$response = curl_exec($ch);
	$httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
	curl_close($ch);

	$responseJson = json_decode($response, true, 512, JSON_THROW_ON_ERROR);

	if ($httpCode < 200 || $httpCode >= 300) {
		throw new OAuthError($responseJson['error'] ?? 'unknown_error', $responseJson['error_description'] ?? 'An unknown error occurred.');
	}

	return $responseJson;
}


/**
 * Returns supported types
 *
 * @param string $login_token
 *
 * @return stdClass
 */
function getSupportedTypes($token) {
	$postData = getDefaultPostData('status');

	$postData['supported_types'] = 1;

	$encodedData = json_encode($postData);

	$result = file_get_contents(
		'https://api.ibroadcast.com/s/JSON/',
		false,
		stream_context_create(
			array(
				'http' => array(
					'method' => 'POST',
					'header' => 'Content-Type: application/json' . "\r\n"
						. 'Content-Length: ' . strlen($encodedData) . "\r\n"
						. 'User-Agent: ' . $postData['user_agent'] . "\r\n"
						. 'Authorization: ' . $token['token_type'] . ' ' . $token['access_token'] . "\r\n",
					'content' => $encodedData,
				),
			)
		)
	);

	return json_decode($result);
}

/**
 * Returns a hash map of uploaded files.
 *
 * @param array $token
 *
 * @return stdClass
 */
function getSyncContent($token)
{
	$postData = http_build_query(
		array(
		)
	);

	$result = file_get_contents(
		'https://upload.ibroadcast.com',
		false,
		stream_context_create(
			array(
				'http' => array(
					'method' => 'POST',
					'header' => 'Content-Type: application/x-www-form-urlencoded' . "\r\n"
						. 'Content-Length: ' . strlen($postData) . "\r\n"
						. 'User-Agent: ' . 'php uploader 1.0' . "\r\n"
						. 'Authorization: ' . $token['token_type'] . ' ' . $token['access_token'] . "\r\n",
					'content' => $postData,
				)
			)
		)
	);

	return json_decode($result);
}

/**
 * Uploads the given file to the remote service.
 *
 * @param string $file
 * @param array $token
 *
 * @return bool True, if the file upload was successful
 */
function uploadFile($file, $token) {
	$multipartBoundary = '-----------'.microtime(true);
	$header = 'Content-Type: multipart/form-data; boundary=' . $multipartBoundary . "\r\n"
			. 'User-Agent: ' . 'php uploader 1.0' . "\r\n"
			. 'Authorization: ' . $token['token_type'] . ' ' . $token['access_token'] . "\r\n";

	$file_contents = file_get_contents($file);
	$contentType   = mime_content_type($file);

	$content .= "--".$multipartBoundary."\r\n".
		"Content-Disposition: form-data; name=\"file_path\"\r\n\r\n".
		"$file\r\n";

	$content .= "--".$multipartBoundary."\r\n".
		"Content-Disposition: form-data; name=\"method\"\r\n\r\n".
		"php uploader\r\n";

	$content .=  "--".$multipartBoundary."\r\n".
		"Content-Disposition: form-data; name=\"file\"; filename=\"".basename($file)."\"\r\n".
		"Content-Type: $contentType\r\n\r\n".
		$file_contents."\r\n";

	$content .= "--".$multipartBoundary."--\r\n";

	file_get_contents(
		'https://upload.ibroadcast.com',
		false,
		stream_context_create(
			array(
				'http' => array(
					'method' => 'POST',
					'header' => $header,
					'content' => $content,
				)
			)
		)
	);

	preg_match('{HTTP\/\S*\s(\d{3})}', $http_response_header[0], $match);
	$status = $match[1];

	if ($status == "401") {
		
	}

	return true;
}

/**
 * Returns an array of allowed file extensions.
 *
 * @param stdClass $serviceContent
 *
 * @return array
 */
function getSupportedFormats($serviceContent) {
	$supportedFormat = array();
	foreach ($serviceContent->supported as $key => $format) {
		$supportedFormat[] = str_replace('.', '', $format->extension);
	}
	return $supportedFormat;
}

/**
 * Recursive function for traversing directories.
 *
 * @param string $dirPath
 * @param array $files
 */
function readFiles($dirPath, &$files) {
	global $supportedFormat, $files;

	$dirContent = scandir($dirPath);

	foreach ($dirContent as $key => $item) {
		// Skipp hidden files and parent folder references
		if (in_array($item, array('.','..')) || $item[0] === '.') {
			continue;
		}

		$itemPath = $dirPath . '/' . $item;

		if (is_dir($itemPath)) {
			readFiles($itemPath, $files);
		} else if (is_file($itemPath)) {
			$fileInfo = explode('.', $item);

			if (count($fileInfo) > 1 && in_array(end($fileInfo), $supportedFormat)) {
				$files[] = $itemPath;
			}
		}
	}
}

exit(0);

File: python/ibroadcast-uploader.py
#!/usr/bin/env python

import json
import glob
import os
import hashlib
import sys
import traceback
import time
import io

import requests
import segno

sys.tracebacklimit = 0

def get_input(inp):
    if sys.version_info >= (3, 0):
        return input(inp)
    else:
        return raw_input(inp)

class OAuthError(Exception):
    def __init__(self, code, message):
        super().__init__(message)
        self.code = code

    def __reduce__(self):
        return (OAuthError, (self.code, self.message))

class ServerError(Exception):
    pass

class ValueError(Exception):
    pass

class Uploader(object):
    """
    Class for uploading content to iBroadcast.
    """

    VERSION = '0.4'
    CLIENT = 'python 3 uploader script'
    DEVICE_NAME = 'python 3 uploader script'
    USER_AGENT = 'python 3 uploader script 0.4'
    CLIENT_ID = 'de4ce836a9fb11f0bc7fb49691aa2236'


    def __init__(self):

        # Initialize our variables that each function will set.
        self.token = None
        self.user = None

        self.supported = None
        self.files = None
        self.md5 = None

    def loadToken(self):
        self.token = None
        try:
            with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'ibroadcast-uploader.json')) as f:
                j = json.load(f)
                self.token = j['token']
        except:
            pass

    def saveToken(self):
        try:
            with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'ibroadcast-uploader.json'), 'w') as f:
                json.dump({ 'token': self.token }, f)
        except Exception as e:
            print(f'Warning, unable to save token to ibroadcast-uploader.json: {str(e)}')


    def login(self):
        device_code = None

        # Need to refresh token?
        self.refreshTokenIfNecessary()

        # Loop until we have a token
        while self.token is None:
            # Get device code if we don't have one yet
            if device_code is None:
                try:
                    device_code = self.oauthDeviceCode()
                    device_code['expires_at'] = time.time() + device_code['expires_in']

                    # Generate and show barcode and URL
                    qrcode = segno.make(device_code['verification_uri_complete'], error='h')
                    buffer = io.StringIO()
                    qrcode.terminal(out=buffer)
                    print(buffer.getvalue())

                    # Visit URL
                    print(f'To authorize, scan the QR code or enter code {device_code['user_code']} at: {device_code['verification_uri']}')

                    print('\nWaiting for authorization...')
                except Exception as e:
                    print(f'Unable to get device code: {str(e)}')
                    return False
            
            # Check device code expiry
            if device_code['expires_at'] <= time.time():
                print('Device code timed out!')
                # Remove it and start over
                device_code = None
                continue
            
            
            # Poll for token
            try:
                self.token = self.oauthToken(device_code['device_code'])
                self.token['expires_at'] = time.time() + self.token['expires_in']
            except OAuthError as e:
                if e.code == 'authorization_pending':
                    time.sleep(device_code['interval'])
                    continue
                    
                print(f'Authorization error: {str(e)}')
                return False
            
            # Authorization successful! Stop looping
            break

        return True

    def process(self):
        # Get supported types and account info
        try:
            self.get_supported_types()
        except ValueError as e:
            print('Unable to fetch account info: %s' % e)
            return

        self.load_files()

        if self.confirm():
            self.upload()

    def oauthDeviceCode(self):
        """
        Gets a device code

        Raises:
            OAuthError

        """
        print('Getting device code...')
        
        query_params = {
            'client_id' : self.CLIENT_ID,
            'scope' : ' '.join(['user.account:read', 'user.upload'])
        }
        response = requests.get(
            'https://oauth.ibroadcast.com/device/code',
            params=query_params,
            headers={'User-Agent': self.USER_AGENT}
        )

        response_json = response.json()

        if not response.ok:
            raise OAuthError(response_json['error'], response_json['error_description'])

        return response_json

    def oauthToken(self, code):
        """
        Gets a token given a device code

        Raises:
            OAuthError

        """
        # print('Getting token...')
        
        body = {
            'client_id' : self.CLIENT_ID,
            'grant_type' : 'device_code',
            'device_code' : code
        }
        response = requests.post(
            'https://oauth.ibroadcast.com/token',
            data=body,
            headers={'User-Agent': self.USER_AGENT}
        )

        response_json = response.json()

        if not response.ok:
            raise OAuthError(response_json['error'], response_json['error_description'])

        return response_json

    def refreshToken(self, refresh_token):
        """
        Refreshes a token

        Raises:
            OAuthError

        """
        print('Refreshing token...')
        
        body = {
            'client_id' : self.CLIENT_ID,
            'grant_type' : 'refresh_token',
            'refresh_token' : refresh_token
        }
        response = requests.post(
            'https://oauth.ibroadcast.com/token',
            data=body,
            headers={'User-Agent': self.USER_AGENT}
        )

        response_json = response.json()

        if not response.ok:
            raise OAuthError(response_json['error'], response_json['error_description'])

        return response_json

    def refreshTokenIfNecessary(self):
        if self.token is None:
            return
        
        if self.token['expires_at'] <= time.time():
            try:
                self.token = self.refreshToken(self.token['refresh_token'])
                self.token['expires_at'] = time.time() + self.token['expires_in']
            except OAuthError as e:
                print(f'Authorization error, please log in again: {str(e)}')
                self.token = None

    def get_supported_types(self):
        """
        Get supported file types

        Raises:
            ValueError on invalid login

        """
        print('Fetching account info...')
        # Build a request object.
        post_data = json.dumps({
            'mode' : 'status',
            'supported_types': 1,
            'version': self.VERSION,
            'client': self.CLIENT,
            'device_name' : self.DEVICE_NAME,
            'user_agent' : self.USER_AGENT
        })
        response = requests.post(
            "https://api.ibroadcast.com/s/JSON/",
            data=post_data,
            headers={'Content-Type': 'application/json', 'User-Agent': self.USER_AGENT, 'Authorization': f'{self.token['token_type']} {self.token['access_token']}'}
        )

        if not response.ok:
            raise ServerError('Server returned bad status: ',
                             response.status_code)

        jsoned = response.json()

        if 'user' not in jsoned:
            raise ValueError(jsoned.message)

        print('Account info fetched')

        self.supported = []
        self.files = []

        for filetype in jsoned['supported']:
             self.supported.append(filetype['extension'])

    def load_files(self, directory=None):
        """
        Load all files in the directory that match the supported extension list.

        directory defaults to present working directory.

        raises:
            ValueError if supported is not yet set.
        """
        if self.supported is None:
            raise ValueError('Supported not yet set - have you logged in yet?')

        if not directory:
            directory = os.getcwd()

        for full_filename in glob.glob(os.path.join(directory, '*')):
            filename = os.path.basename(full_filename)
            # Skip hidden files.
            if filename.startswith('.'):
                continue

            # Make sure it's a supported extension.
            dummy, ext = os.path.splitext(full_filename)
            if ext in self.supported:
                self.files.append(full_filename)

            # Recurse into subdirectories.
            # XXX Symlinks may cause... issues.
            if os.path.isdir(full_filename):
                self.load_files(full_filename)

    def confirm(self):
        """
        Presents a dialog for the user to either list all files, or just upload.
        """
        print("Found %s files.  Press 'L' to list, or 'U' to start the " \
              "upload." % len(self.files))
        response = get_input('--> ')

        print()
        if response == 'L'.upper():
            print('Listing found, supported files')
            for filename in self.files:
                print(' - ', filename)
            print()
            print("Press 'U' to start the upload if this looks reasonable.")
            response = get_input('--> ')
        if response == 'U'.upper():
            print('Starting upload.')
            return True

        print('Aborting')
        return False

    def __load_md5(self):
        """
        Reach out to iBroadcast and get an md5.
        """
        # Send our request.
        response = requests.post(
            "https://upload.ibroadcast.com",
            headers={'Authorization': f'{self.token['token_type']} {self.token['access_token']}'}
        )

        if not response.ok:
            raise ServerError('Server returned bad status: ',
                             response.status_code)

        jsoned = response.json()

        self.md5 = jsoned['md5']

    def calcmd5(self, filePath="."):
        with open(filePath, 'rb') as fh:
            m = hashlib.md5()
            while True:
                data = fh.read(8192)
                if not data:
                    break
                m.update(data)
        return m.hexdigest()

    def upload(self):
        """
        Go and perform an upload of any files that haven't yet been uploaded
        """
        self.__load_md5()

        for filename in self.files:

            print('Uploading ', filename)

            # Get an md5 of the file contents and compare it to whats up
            # there already
            file_md5 = self.calcmd5(filename)

            if file_md5 in self.md5:
                print('Skipping - already uploaded.')
                continue
            upload_file = open(filename, 'rb')

            file_data = {
                'file': upload_file,
            }

            post_data = {
                'file_path' : filename,
                'method': self.CLIENT,
            }

            response = requests.post(
                "https://upload.ibroadcast.com",
                post_data,
                files=file_data,
                headers={'Authorization': f'{self.token['token_type']} {self.token['access_token']}'}
            )

            ## refresh and retry one time
            if response.status_code == 401:
                self.refreshTokenIfNecessary()
                response = requests.post(
                    "https://upload.ibroadcast.com",
                    post_data,
                    files=file_data,
                    headers={'Authorization': f'{self.token['token_type']} {self.token['access_token']}'}
                )

            upload_file.close()

            if not response.ok:
                raise ServerError('Server returned bad status: ',
                    response.status_code)
            jsoned = response.json()
            result = jsoned['result']

            if result is False:
                raise ValueError('File upload failed.')
        print('Done')

if __name__ == '__main__':
    uploader = Uploader()

    # Load token from local storate if possible
    uploader.loadToken()

    # Login or refresh token if necessary
    if uploader.login():
        # Save token to local storage
        uploader.saveToken()

        # Ready to start processing files
        uploader.process()
